# Exploring Hacker News Posts
In this project, we are going to analyze data from the website [Hacker News](https://news.ycombinator.com/) which is a site started by the startup incubator [Y Combinator](https://www.ycombinator.com/) where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit.
The original dataset in this project can be found [here](https://www.kaggle.com/hacker-news/hacker-news-posts). 
<br> However, it has been reduced from almost 300,000 rows to approximatly 20,000 by removing all submissions that did not receive any comment, and then randomly sampling from the remaining submissions.
<br> Below are descriptions of the columns:
- id: The unique identifier from Hacker News for the post
- title: The title of the post
- url: The URL that the posts links to, if it the post has a URL
- num_points: The number of points the post acquired, calculated as the - total number of upvotes minus the total number of downvotes
- num_comments: The number of comments that were made on the post
- author: The username of the person who submitted the post
- created_at: The date and time at which the post was submitted

In this analyze, we are going to focus on posts whose titles begin with either Ask HN or Show HN.
<br> Users submit **Ask HN** posts to ask the Hacker News community a specific question. Below are a couple examples:

>Ask HN: How to improve my personal website?
<br> Ask HN: Am I the only one outraged by Twitter shutting down share counts?
<br> Ask HN: Aby recent changes to CSS that broke mobile?

Likewise, users submit **Show HN** posts to show the Hacker News community a project, product, or just generally something interesting. Below are a couple of examples:

>Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development        Platform'
<br> Show HN: Something pointless I made
<br> Show HN: Shanhu.io, a programming playground powered by e8vm

This analyze will compare these two types of posts to detemine :
- Whether Ask HN or Show HN receive more comments on average?
- Whether posts created at a particular time receive more comments on average?

First, let's start by importing the libraries we need and reading the data set into a list of lists.

In [1]:
import csv
f = open('hacker_news.csv')
reader = csv.reader(f)
hn = list(reader)
for row in hn[:5] :
    print(row)
    print('\n')

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']




In [2]:
headers = hn[0]
hn = hn[1:]
print(headers)
print('\n')
for row in hn[:5] :
    print(row)
    print('\n')

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52']


['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30']


['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']


['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']


['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']




## Filtering Ask HN, Show HN and Other posts
Now let's filter our data and separate the **Ask HN**, **Show HN**, and the other posts by creating new lists of lists that will only contain those posts.
<br> We will also determine how many of each type of post are contained in our data for analyzis

In [3]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn :
    title = (row[1].lower())
    if title.startswith('ask hn') :
        ask_posts.append(row)
    elif title.startswith('show hn') :
        show_posts.append(row)
    else:
        other_posts.append(row)

print('There are '+ str(len(ask_posts))+ ' Ask HN posts')
print('There are '+ str(len(show_posts))+ ' Show HN posts')
print('There are '+ str(len(other_posts))+ ' other posts')

There are 1744 Ask HN posts
There are 1162 Show HN posts
There are 17194 other posts


An important thing to notice is that there are far more **Ask HN** posts than **Show HN** posts.

## Computing the Average number of comments for Show and Ask posts

In [4]:
total_ask_comments = 0
for row in ask_posts :
    n_comments = int(row[4])
    total_ask_comments += n_comments
total_show_comments = 0
for row in show_posts :
    n_comments = int(row[4])
    total_show_comments += n_comments

avg_ask_comments = total_ask_comments / len(ask_posts)
avg_show_comments = total_show_comments / len(show_posts)
print('The average number of comments for a Ask HN post is', avg_ask_comments)
print('The average number of comments for a Show HN post is', avg_show_comments)

The average number of comments for a Ask HN post is 14.038417431192661
The average number of comments for a Show HN post is 10.31669535283993


Based on our findings, we can conclude that on average, a **Ask HN** post receives more comments than a **Show HN** post.
<br> And like we noticed earlier, there are more **Ask HN** posts(about 1744) than **Show HN** posts(about 1162).
<br> Because **Ask HN** posts receive more comments on average.
We are going to focus on the **Ask HN** posts for the remaining parts of our analyzis

## Computing Ask Hn post created per Hours
Now let's look at the time *Ask HN* posts where created at, to determine whether posts created at a certain time are more likely to attract comments. To achieve this, we are going to:
- Calculate the amount of *Ask HN* posts created in each hour of the day, along with the number of comments received.
- Calculate the average number of comments *Ask HN* posts receive by hour created.

In [5]:
import datetime as dt
result_list = []
for row in ask_posts :
    result_list.append([row[6], int(row[4])])
print(result_list[:5])

counts_by_hour = {}
comments_by_hour ={}
for row in result_list :
    n_comments = row[1]
    date_created = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M')
    hour = date_created.strftime('%H')
    if hour not in counts_by_hour :
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = n_comments
    else :
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += n_comments

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]


In the above step, we comptuted the number of comments made for each hour and the number of posts created each hour.
Now, we are going to compute the average number of comments that posts receive for each hour 

In [6]:
avg_by_hour = []
for hour in comments_by_hour :
    avg_comments_by_hour = comments_by_hour[hour] / counts_by_hour[hour]
    avg_by_hour.append([hour, avg_comments_by_hour])

print(avg_by_hour[:5])

[['06', 9.022727272727273], ['19', 10.8], ['13', 14.741176470588234], ['22', 6.746478873239437], ['04', 7.170212765957447]]


In order to make it easier to identify the hours with the most comments on average, let's sort the list of lists and print the five highest values in a format that's easier to read.

In [7]:
swap_avg_by_hour= []
for row in avg_by_hour :
    swap_avg_by_hour.append([row[1], row[0]])
print(swap_avg_by_hour)

[[9.022727272727273, '06'], [10.8, '19'], [14.741176470588234, '13'], [6.746478873239437, '22'], [7.170212765957447, '04'], [11.46, '17'], [7.985294117647059, '23'], [38.5948275862069, '15'], [10.08695652173913, '05'], [16.009174311926607, '21'], [11.383333333333333, '01'], [13.233644859813085, '14'], [8.127272727272727, '00'], [11.051724137931034, '11'], [7.852941176470588, '07'], [5.5777777777777775, '09'], [16.796296296296298, '16'], [23.810344827586206, '02'], [10.25, '08'], [21.525, '20'], [9.41095890410959, '12'], [13.440677966101696, '10'], [13.20183486238532, '18'], [7.796296296296297, '03']]


In [8]:
sorted_swap = sorted(swap_avg_by_hour, reverse = True)
print("Top 5 Hours for Ask Posts Comments")
for row in sorted_swap[:5] :
    hour = dt.datetime.strptime(row[1], '%H')
    hour = hour.strftime('%H:%M')
    average_comments = row[0]
    print('{}: {:.2f} average comments per post.'.format(hour, average_comments))


Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post.
02:00: 23.81 average comments per post.
20:00: 21.52 average comments per post.
16:00: 16.80 average comments per post.
21:00: 16.01 average comments per post.


## Conclusion
Our analzsis shows that, for a post to have the highest chance of receiving a comment, the post has to be created at **3pm**.
<br> It is imposrtant to notice that the hours refered to in this analysis are in *Eastern Standard Time* (EST) or *US/Eastern timezone*.
<br> We would need to convert them to account for different timezone/cities/countries/.
For example, because I live in Dallas, Tx which is in *Central Daylight* Time or *US/Central*.
For a post to have the highest chance of receiving a comment, a person in Dallas, Tx would need to post it at **2pm**

In [9]:
from pytz import timezone
time = dt.datetime.strptime('15:00','%H:%M')
time_w_zone = time.replace(tzinfo=timezone('US/Eastern'))
time_in_dallas = time_w_zone.astimezone(timezone('US/Central'))
print(time_in_dallas.strftime('%H:%M'))

14:05
